File: src/csharp/Problems/IsAnagram.cs
//https://leetcode.com/problems/add-two-integers/

namespace LeetCode.Problems;

public sealed class IsAnagram : ProblemBase
{
    [Theory]
    [ClassData(typeof(IsAnagram))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.Param("anagram").Param("nagaram").Result(true))
          .Add(it => it.Param("anagram").Param("nagara").Result(false))
          .Add(it => it.Param("rat").Param("car").Result(false));

    private bool Solution(string s, string t)
    {
        if (s.Length != t.Length)
        {
            return false;
        }

        var map = new Dictionary<char, int>();
        for (var i = 0; i < s.Length; i++)
        {
            if (!map.ContainsKey(s[i]))
            {
                map.Add(s[i], 0);
            }

            map[s[i]]++;

            if (!map.ContainsKey(t[i]))
            {
                map.Add(t[i], 0);
            }

            map[t[i]]--;
        }

        return map.All(it => it.Value == 0);
    }
}

File: src/go/problems/IsAnagram.go
// https://leetcode.com/problems/running-sum-of-1d-array/

package problems

func IsAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	set := map[byte]int{}
	length := len(s)
	for i := 0; i < length; i++ {
		set[s[i]]++
		set[t[i]]--
	}

	for _, v := range set {
		if v != 0 {
			return false
		}
	}

	return true
}
